examples/android/clientcache/app/src/main/java/io/grpc/clientcacheexample/SafeMethodCachingInterceptor.java

```diff
@@ -0,0 +1,297 @@
+                        cacheResponse = false;
+                        break;
+                      } else if (directive.equalsIgnoreCase("no-transform")) {
+                        cacheResponse = false;
+                        break;
+                      } else if (directive.toLowerCase().startsWith("max-age")) {
+                        String[] parts = directive.split("=");
+                        if (parts.length == 2) {
+                          try {
+                            maxAge = Integer.parseInt(parts[1]);
+                          } catch (NumberFormatException e) {
+                            continue;
+                          }
+                        }
+                      }
+                    }
+                  }
+                }
+                if (cacheResponse) {
+                  if (maxAge > -1) {
+                    deadline = Deadline.after(maxAge, TimeUnit.SECONDS);
+                  } else {
+                    deadline = Deadline.after(defaultMaxAge, TimeUnit.SECONDS);
+                  }
+                }
+                super.onHeaders(headers);
+              }
+
+              @Override
+              public void onMessage(RespT message) {
+                if (cacheResponse && !deadline.isExpired()) {
+                  Value value = new Value((MessageLite) message, deadline);
+                  internalCache.put(requestKey, value);
+                }
+                super.onMessage(message);
+              }
+
+              @Override
+              public void onClose(Status status, Metadata trailers) {
+                if (cacheOptionsCausedCancellation) {
+                  // UNAVAILABLE is the canonical gRPC mapping for HTTP response code 504 (as used
+                  // by the built-in Android HTTP request cache).
+                  super.onClose(
+                      Status.UNAVAILABLE.withDescription(status.getDescription()), new Metadata());
+                } else {
+                  super.onClose(status, trailers);
+                }
+              }
+            };
+        delegate().start(interceptedListener, headers);
+      }
+
+      @Override
+      public void sendMessage(ReqT message) {
+        boolean noCache = callOptions.getOption(NO_CACHE_CALL_OPTION);
+        boolean onlyIfCached = callOptions.getOption(ONLY_IF_CACHED_CALL_OPTION);
+
+        if (noCache) {
+          if (onlyIfCached) {
+            cacheOptionsCausedCancellation = true;
+            super.cancel("Unsatisfiable Request (no-cache and only-if-cached conflict)", null);
+            return;
+          }
+          cacheResponse = false;
+          super.sendMessage(message);
+          return;
+        }
+
+        // Check the cache
+        requestKey = new Key(fullMethodName, (MessageLite) message);
+        Value cachedResponse = internalCache.get(requestKey);
+        if (cachedResponse != null) {
+          if (cachedResponse.maxAgeDeadline.isExpired()) {
```

Comment:
if there are no requests for a long time, the cache could build up to be pretty big correct?

Reply:
The LruCache used here has a built-in size limit